### ISSUE TITLE
separate branches for bug fixes and enhancements

### ISSUE THREAD

**Issue by shiroyagi** [osmdroid/osmdroid#982]
## Issue Type

[*] Improvement

## Description and/or steps/code to reproduce the problem

When a version of osmdroid is released, work soon starts on developing new features and making improvements for the next version.  At the same time bugs may be discovered in the existing release but both are mixed on the master branch.  This may include API changes and potentially new bugs, so that it can be difficult to find a stable version to use.

I would suggest that following a release, bug fixes are made to a maintenance branch for that release (representing a major version number) and merged across to the master branch where applicable.  Also that breaking API changes (eg. removal or changes to method arguments) only happen on the master branch, representing a major version number change.

Hopefully this would involve minimal additional effort.  By the time a new major release is ready, assuming there are no serious outstanding bugs for the previous version, its maintenance branch can be closed and a new one created.

### Version of osmdroid the issue relates to:

All

**Comment by monsieurtanuki** [comment]
@shiroyagi I'm definitely not an expert on git branches and I have clearly no idea about how the other developers integrate osmdroid into their apps technically (I only know how _I_ do it), how often they do it, or the best practices around that.

As a developer of bug fixes, I'd say that most of the time it involves refactoring, which makes it painful to keep the API without potentially inconvenient changes.

I'd rather be in favor of some kind of _frozen zone_, something like a month after the latest major version delivery, where we developers focus only on making this version bullet-proof (typical manual tests, additional unit tests, ...).
And then keep this very decent version available as the latest official osmdroid release like a dead branch, and keep coding new features and bug fixes in the current branch, with possible unexpected bugs for whoever downloads it. Until we decide to create a new version and polish it during the next _frozen zone_.

**Comment by shiroyagi** [comment]
Yes that sounds good too, I think there's a place for both approaches.  The bullet-proof version could be on a maintenance branch, or a maintenance branch could be subsequently be created from the commit immediately prior to the first breaking API change (but this may not be immediately obvious).  Some bug fixes are simple one-liners, or there maybe a simple mitigation with a full fix requiring more major refactoring.

With the work being done on a voluntary basis, it's really up to those contributing to decide where they wish to spend their effort and how long to maintain the previous version.  I would certainly prefer to spend my time on version I can easily use without coming across other problems. Having an official maintenance branch would give people that option.

I'm no expert on git but like other modern SCMs I think branches are cheap and merges easy.  All this really needs to work is the creation of the branch and a commitment from those merging to it to vet the changes accordingly.

**Comment by spyhunter99** [comment]
how about master is the last release and bug fixes. new features go to develop?

**Comment by Tester2012** [comment]
My 5 cents:

There's always a dev branch used for features and bug fixes.
1-2 weeks before an official release, a new branch checked out from the dev branch is created. It accepts only critical bug fixes. Then, the release branch is rolled out and merged into the dev branch and master branch.
Master branch is always a stable branch that's used neither for dev nor for release. The release branch is merged back into it once the release branch is rolled out.

**Comment by shiroyagi** [comment]
The key for me is the separation of bugs from enhancements for as long as people wish to maintain them.  A general preference for fixing bugs on maintenance then merging to development using whatever branch layout is most convenient.  When I get chance I will try cherry-picking some recent fixes onto the last release.  If there was a suitable branch I could then create a PR.
